server/src/main/java/io/druid/server/coordinator/cost/SegmentCostCache.java

```diff
@@ -0,0 +1,379 @@
+      Comparators.intervalsByStartThenEnd().compare(
+          d1.getInterval(),
+          d2.getInterval()
+      );
+
+  private static final Comparator<Bucket> BUCKET_INTERVAL_COMPARATOR = (b1, b2) ->
+      Comparators.intervalsByStartThenEnd().compare(
+          b1.getInterval(),
+          b2.getInterval()
+      );
+
+  private final List<Bucket> sortedBuckets;
+  private final List<Interval> intervals;
+
+  public SegmentCostCache(List<Bucket> sortedBuckets)
+  {
+    this.sortedBuckets = Preconditions.checkNotNull(sortedBuckets, "buckets should not be null");
+    this.intervals = sortedBuckets.stream().map(Bucket::getInterval).collect(Collectors.toCollection(ArrayList::new));
+    Preconditions.checkArgument(
+        Ordering.from(BUCKET_INTERVAL_COMPARATOR).isOrdered(sortedBuckets),
+        "buckets must be ordered by interval"
+    );
+  }
+
+  public double cost(DataSegment segment)
+  {
+    double cost = 0.0;
+    int index = Collections.binarySearch(intervals, segment.getInterval(), Comparators.intervalsByStartThenEnd());
+    index = (index >= 0) ? index : -index - 1;
+
+    ListIterator<Bucket> it = sortedBuckets.listIterator(index);
+    while (it.hasNext()) {
+      Bucket bucket = it.next();
+      if (!bucket.inCalculationInterval(segment)) {
+        break;
+      }
+      cost += bucket.cost(segment);
+    }
+
+    it = sortedBuckets.listIterator(index);
+    while (it.hasPrevious()) {
+      Bucket bucket = it.previous();
+      if (!bucket.inCalculationInterval(segment)) {
+        break;
+      }
+      cost += bucket.cost(segment);
+    }
+
+    return cost;
+  }
+
+  public static Builder builder()
+  {
+    return new Builder();
+  }
+
+  public static class Builder
+  {
+    private NavigableMap<Interval, Bucket.Builder> buckets = new TreeMap<>(Comparators.intervalsByStartThenEnd());
+
+    public void addSegment(DataSegment segment)
+    {
+      Bucket.Builder builder = buckets.computeIfAbsent(toBucketInterval(segment), Bucket::builder);
+      builder.addSegment(segment);
+    }
+
+    public void removeSegement(DataSegment segment)
+    {
+      Interval interval = toBucketInterval(segment);
+      Bucket.Builder builder = buckets.get(interval);
+      if (builder != null) {
+        builder.removeSegment(segment);
+        if (builder.size() == 0) {
+          buckets.remove(interval);
+        }
+      }
+    }
+
+    public int size()
+    {
+      return buckets.size();
+    }
+
+    public SegmentCostCache build()
+    {
+      return new SegmentCostCache(
+          buckets
+              .entrySet()
+              .stream()
+              .map(entry -> entry.getValue().build())
+              .collect(Collectors.toCollection(ArrayList::new))
+      );
+    }
+
+    private Interval toBucketInterval(DataSegment segment)
+    {
+      long start = segment.getInterval().getStartMillis() - (segment.getInterval().getStartMillis() / BUCKET_INTERVAL);
```

Comment:
it's a dummy bug, should be 
` long start = segment.getInterval().getStartMillis() - (segment.getInterval().getStartMillis() % BUCKET_INTERVAL);`
btw, it has huge impact on the performance